x-pack/auditbeat/module/system/login/utmp.go

```diff
@@ -0,0 +1,571 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build linux,cgo
+
+package login
+
+// #include <utmp.h>
+// #include <stdlib.h>
+import "C"
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/user"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strconv"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/auditbeat/datastore"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const (
+	bucketKeyFileRecords   = "file_records"
+	bucketKeyLoginSessions = "login_sessions"
+)
+
+// Inode represents a file's inode on Linux.
+type Inode uint64
+
+// FileRecord represents a UTMP file at a point in time.
+type FileRecord struct {
+	Inode    Inode
+	Size     int64
+	LastUtmp Utmp
+}
+
+// Utmp contains data from the C utmp struct.
+type Utmp struct {
+	UtType   int
+	UtPid    int
+	UtLine   string
+	UtUser   string
+	UtHost   string
+	UtTv     time.Time
+	UtAddrV6 [4]uint32
+}
+
+func newUtmp(utmpC *C.struct_utmp) Utmp {
+	// See utmp(5) for the utmp struct fields.
+	return Utmp{
+		UtType:   int(utmpC.ut_type),
+		UtPid:    int(utmpC.ut_pid),
+		UtLine:   C.GoString(&utmpC.ut_line[0]),
+		UtUser:   C.GoString(&utmpC.ut_user[0]),
+		UtHost:   C.GoString(&utmpC.ut_host[0]),
+		UtTv:     time.Unix(int64(utmpC.ut_tv.tv_sec), int64(utmpC.ut_tv.tv_usec)*1000),
+		UtAddrV6: [4]uint32{uint32(utmpC.ut_addr_v6[0]), uint32(utmpC.ut_addr_v6[1]), uint32(utmpC.ut_addr_v6[2]), uint32(utmpC.ut_addr_v6[3])},
+	}
+}
+
+// UtmpFileReader can read a UTMP formatted file (usually /var/log/wtmp).
+type UtmpFileReader struct {
+	log           *logp.Logger
+	bucket        datastore.Bucket
+	filePattern   string
+	fileRecords   map[Inode]FileRecord
+	loginSessions map[string]LoginRecord
+}
+
+// NewUtmpFileReader creates and initializes a new UTMP file reader.
+func NewUtmpFileReader(log *logp.Logger, bucket datastore.Bucket, filePattern string) (*UtmpFileReader, error) {
+	r := &UtmpFileReader{
+		log:           log,
+		bucket:        bucket,
+		filePattern:   filePattern,
+		fileRecords:   make(map[Inode]FileRecord),
+		loginSessions: make(map[string]LoginRecord),
+	}
+
+	// Load state (file records, tty mapping) from disk
+	err := r.restoreStateFromDisk()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to restore state from disk")
+	}
+
+	return r, nil
+}
+
+// Close performs any cleanup tasks when the UTMP reader is done.
+func (r *UtmpFileReader) Close() error {
+	err := r.bucket.Close()
+	if err != nil {
+		return errors.Wrap(err, "error closing bucket")
+	}
+
+	return nil
+}
+
+// ReadNew returns any new UTMP entries in any files matching the configured pattern.
+func (r *UtmpFileReader) ReadNew() ([]LoginRecord, error) {
```

Comment:
How big do we expect these files to be? Looks like this function will load all available logs into memory on first run.

Reply:
So far I've assumed they're usually quite small, but that might not be true in all cases. The `/etc/logrotate.conf` on my Ubuntu VM rotates it monthly. On that VM that's getting a number of daily logins/logouts, they're <100K. On a real-world machine with a lot of logins, I assume it could be more, I wonder how many records it would take to be significant.

Nevertheless, I guess we could use a channel between `Fetch` and `ReadNew` and output records as events as they come.

Reply:
This is more or less what filebeat does as well. Have one go-routine per file and forward event by event. That is, the module/input is more or less active all the time. Not sure how well this fits the 'Fetch' semantics in the metricbeat framework. But a bounded channel can indeed help keeping memory usage in place. Given the complexity this introduces I wonder if that's really a problem.

Reply:
I've changed to using channels now, and storing offsets of the files (instead of re-reading until the last read record). So hopefully that will make it scalable even if the files are very large.